plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/preferences/RunOptionsDefaultsComponent.java

```diff
@@ -0,0 +1,312 @@
+import com.google.cloud.tools.eclipse.dataflow.ui.DataflowUiPlugin;
+import com.google.cloud.tools.eclipse.dataflow.ui.page.MessageTarget;
+import com.google.cloud.tools.eclipse.dataflow.ui.util.ButtonFactory;
+import com.google.cloud.tools.eclipse.dataflow.ui.util.DisplayExecutor;
+import com.google.cloud.tools.eclipse.dataflow.ui.util.SelectFirstMatchingPrefixListener;
+import com.google.cloud.tools.eclipse.dataflow.ui.util.SelectFirstMatchingPrefixListener.OnCompleteListener;
+import com.google.common.base.Strings;
+
+import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.events.FocusAdapter;
+import org.eclipse.swt.events.FocusEvent;
+import org.eclipse.swt.events.ModifyListener;
+import org.eclipse.swt.events.SelectionAdapter;
+import org.eclipse.swt.events.SelectionEvent;
+import org.eclipse.swt.graphics.Point;
+import org.eclipse.swt.layout.GridData;
+import org.eclipse.swt.widgets.Button;
+import org.eclipse.swt.widgets.Combo;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Control;
+import org.eclipse.swt.widgets.Label;
+import org.eclipse.swt.widgets.Text;
+import org.eclipse.ui.statushandlers.StatusManager;
+
+import java.util.SortedSet;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+
+/**
+ * Collects default run options for Dataflow Pipelines and provides means to create and modify them.
+ */
+public class RunOptionsDefaultsComponent {
+  private static final int PROJECT_INPUT_SPENT_COLUMNS = 1;
+  private static final int STAGING_LOCATION_SPENT_COLUMNS = 2;
+
+  private static final long VERIFY_LOCATION_DELAY_MS = 250L;
+
+  private final GcsDataflowProjectClient client;
+  private final DisplayExecutor executor;
+
+  private final MessageTarget messageTarget;
+
+  private final Composite target;
+  private final Text projectInput;
+  private final Combo stagingLocationInput;
+  private final Button createButton;
+
+  private VerifyStagingLocationJob verifyJob;
+  private SelectFirstMatchingPrefixListener completionListener;
+
+  public RunOptionsDefaultsComponent(
+      Composite target, int columns, MessageTarget messageTarget, DataflowPreferences preferences) {
+    checkArgument(columns >= 3, "DefaultRunOptions must be in a Grid with at least 3 columns");
+    this.target = target;
+    this.messageTarget = messageTarget;
+    this.executor = DisplayExecutor.create(target.getDisplay());
+    this.client = setupGcsClient();
+
+    Label projectInputLabel = new Label(target, SWT.NULL);
+    projectInput = new Text(target, SWT.SINGLE | SWT.BORDER);
+
+    Label comboLabel = new Label(target, SWT.NULL);
+    stagingLocationInput = new Combo(target, SWT.DROP_DOWN);
+    createButton = ButtonFactory.newPushButton(target, "&Create");
+
+    // Initialize the Default Project, which is used to populate the Staging Location field
+    String project = preferences.getDefaultProject();
```

Comment:
We should think about how we might combine project selectors. We have this for App Engine too.